hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowMetadataTableHistoryProcedure.scala

```diff
@@ -0,0 +1,293 @@
+
+    val metadataTableBasePath = HoodieTableMetadata.getMetadataTableBasePath(dataTableBasePath)
+    val metadataMetaClient = try {
+      Some(HoodieTableMetaClient.builder()
+        .setConf(HadoopFSUtils.getStorageConfWithCopy(jsc.hadoopConfiguration()))
+        .setBasePath(metadataTableBasePath)
+        .setLoadActiveTimelineOnLoad(false)
+        .build())
+    } catch {
+      case _: Exception =>
+        None
+    }
+    metadataMetaClient match {
+      case Some(mdMetaClient) => getTimelineEntries(metaClient, mdMetaClient, limit, showArchived, filter, startTime, endTime)
+      case None => Seq.empty[Row]
+    }
+  }
+
+  override def build: Procedure = new ShowMetadataTableHistoryProcedure()
+
+  private def getTimelineEntries(
+                                  dataMetaClient: HoodieTableMetaClient,
+                                  metadataMetaClient: HoodieTableMetaClient,
+                                  limit: Int,
+                                  showArchived: Boolean,
+                                  filter: String,
+                                  startTime: String,
+                                  endTime: String): Seq[Row] = {
+
+    val dataInstantInfoMap = buildInstantInfoFromTimeline(dataMetaClient)
+    val metadataInstantInfoMap = buildInstantInfoFromTimeline(metadataMetaClient)
+
+    val allInstantTimes = collection.mutable.Set[String]()
+
+    val dataActiveTimeline = dataMetaClient.getActiveTimeline
+    val dataArchivedTimeline = if (showArchived) Some(dataMetaClient.getArchivedTimeline) else None
+
+    val metadataActiveTimeline = metadataMetaClient.getActiveTimeline
+    val metadataArchivedTimeline = if (showArchived) Some(metadataMetaClient.getArchivedTimeline) else None
+
+    allInstantTimes ++= dataActiveTimeline.getInstants.iterator().asScala.map(_.requestedTime())
+    if (dataArchivedTimeline.isDefined) {
+      allInstantTimes ++= dataArchivedTimeline.get.getInstants.iterator().asScala.map(_.requestedTime())
+    }
+
+    allInstantTimes ++= metadataActiveTimeline.getInstants.iterator().asScala.map(_.requestedTime())
+    if (metadataArchivedTimeline.isDefined) {
+      allInstantTimes ++= metadataArchivedTimeline.get.getInstants.iterator().asScala.map(_.requestedTime())
+    }
+
+    val filteredInstantTimes = if (startTime.nonEmpty && endTime.nonEmpty) {
+      allInstantTimes.filter { instantTime =>
+        val timeMatches = (instantTime >= startTime && instantTime <= endTime)
+        timeMatches
+      }
+    } else {
+      allInstantTimes
+    }
+
+    val sortedInstantTimes = filteredInstantTimes.toSeq.sorted(Ordering[String].reverse)
+
+    val entries = sortedInstantTimes.map { instantTime =>
+      createTimelineEntry(
+        instantTime,
+        dataActiveTimeline,
+        dataArchivedTimeline,
+        metadataActiveTimeline,
+        metadataArchivedTimeline,
+        dataInstantInfoMap,
+        metadataInstantInfoMap
+      )
+    }
```

Comment:
I think this is kinda verbose, maybe 

```suggestion
    val entries =
  allInstantTimes
    .filter(t => startTime.nonEmpty && endTime.nonEmpty || (t >= startTime && t <= endTime))
    .toSeq
    .sorted(Ordering.String.reverse)
    .map { instantTime =>
      createTimelineEntry(
        instantTime,
        dataActiveTimeline,
        dataArchivedTimeline,
        metadataActiveTimeline,
        metadataArchivedTimeline,
        dataInstantInfoMap,
        metadataInstantInfoMap
      )
    }
```